jni/src/faiss_index_service.cpp

```diff
@@ -137,28 +218,35 @@ void BinaryIndexService::createIndex(
         throw std::runtime_error("Number of IDs does not match number of vectors");
     }
 
-    std::unique_ptr<faiss::IndexBinary> indexWriter(faissMethods->indexBinaryFactory(dim, indexDescription.c_str()));
-
     // Set thread count if it is passed in as a parameter. Setting this variable will only impact the current thread
     if(threadCount != 0) {
         omp_set_num_threads(threadCount);
     }
 
-    // Add extra parameters that cant be configured with the index factory
-    SetExtraParameters<faiss::IndexBinary, faiss::IndexBinaryIVF, faiss::IndexBinaryHNSW>(jniUtil, env, parameters, indexWriter.get());
-
-    // Check that the index does not need to be trained
-    if(!indexWriter->is_trained) {
-        throw std::runtime_error("Index is not trained");
-    }
+    faiss::IndexBinaryIDMap * idMap = reinterpret_cast<faiss::IndexBinaryIDMap *> (idMapAddress);
 
     // Add vectors
-    std::unique_ptr<faiss::IndexBinaryIDMap> idMap(faissMethods->indexBinaryIdMap(indexWriter.get()));
     idMap->add_with_ids(numVectors, inputVectors->data(), ids.data());
+}
+
+void BinaryIndexService::writeIndex(
+        std::string indexPath,
+        jlong idMapAddress
+    ) {
+
+    std::unique_ptr<faiss::IndexBinaryIDMap> idMap (reinterpret_cast<faiss::IndexBinaryIDMap *> (idMapAddress));
+
+    try {
+        // Write the index to disk
+        faissMethods->writeIndexBinary(idMap.get(), indexPath.c_str());
+    } catch(std::exception &e) {
+        delete idMap->index;
+        throw std::runtime_error("Failed to write index to disk");
+    }
 
-    // Write the index to disk
-    faissMethods->writeIndexBinary(idMap.get(), indexPath.c_str());
+    // Free the memory used by the index
+    delete idMap->index;
```

Comment:
Can we set owns field to true in https://github.com/facebookresearch/faiss/blob/main/faiss/IndexIDMap.h so we dont have to do this?

Reply:
+1, if it works then yes we should do this. But lets ensure all the internal indices are getting freed, otherwise it will be a memory leak.

Reply:
@navneet1v Seems like the [destructor deletes it](https://github.com/facebookresearch/faiss/blob/33c0ba5d002a7cd9761513f06ecc9822079d4a2f/faiss/IndexIDMap.cpp#L211)

Reply:
@shatejas so what you are saying is we are able to set the own_fields to true? and if it is set to true I can see the index is getting deleted.

Reply:
We are able to set it to true. The link is the traced code path and seems like its getting deleted. 

Additional verifications will need an component test which is not currently present, current unit test will not be able to test the delete as its mocked currently. I can revert back to a manual delete if component test is a blocker

Reply:
But if that is the case, if I see this code which builds the index it is already setting the own_fields to true. ref: https://github.com/facebookresearch/faiss/blob/33c0ba5d002a7cd9761513f06ecc9822079d4a2f/faiss/index_factory.cpp#L746-L760 , so we setting that to true should not impact.

and in that case the internal index should have been deleted. But this check was explicitly added because the memory was not getting free when tested in memory constraint env. I would recommend keeping this check as it ensures memory is getting free, and do a separate benchmarks to provide setting own fields is working or not.


If there is any code piece that is reverting this value to value please share. I didn't find any thing. Because I do remember me not freeing the memory caused the leak in my tests and same happened with Andrew too. 


Please let me know I am missing anything. This is what based on my experience working on this code I have seen

Reply:
I think we can keep it. So what we are doing if I see the code correctly, we are using a factory to build the HNSW index, which sets own_field=true. But after that we are putting that HNSW index in the IDMap index there we are not setting own_fields=true. With your change I think it will work.